velox/functions/sparksql/ToJson.h

```diff
@@ -0,0 +1,386 @@
+  auto arrayView = input.castTo<Array<Any>>();
+  result.append("[");
+  for (int i = 0; i < arrayView.size(); i++) {
+    if (i > 0) {
+      result.append(",");
+    }
+    if (arrayView[i].has_value()) {
+      auto element = arrayView[i].value();
+      VELOX_DYNAMIC_TYPE_DISPATCH_ALL(
+          toJson, element.kind(), element, result, options);
+    } else {
+      result.append("null");
+    }
+  }
+  result.append("]");
+}
+
+template <>
+void toJson<TypeKind::MAP>(
+    const exec::GenericView& input,
+    std::string& result,
+    const JsonOptions& options) {
+  auto mapView = input.castTo<Map<Any, Any>>();
+  result.append("{");
+  for (int i = 0; i < mapView.size(); i++) {
+    if (i > 0) {
+      result.append(",");
+    }
+    auto element = mapView.atIndex(i);
+    auto key = element.first;
+    auto value = element.second;
+    bool isKeyVarchar = key.kind() == TypeKind::VARCHAR;
+    if (!isKeyVarchar) {
+      result.append("\"");
+    }
+    VELOX_DYNAMIC_TYPE_DISPATCH_ALL(toJson, key.kind(), key, result, options);
+    result.append(isKeyVarchar ? ":" : "\":");
+    if (value.has_value()) {
+      VELOX_DYNAMIC_TYPE_DISPATCH_ALL(
+          toJson, value.value().kind(), value.value(), result, options);
+    } else {
+      result.append("null");
+    }
+  }
+  result.append("}");
+}
+
+template <typename T>
+struct ToJsonFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const std::vector<TypePtr>& inputTypes,
+      const core::QueryConfig& config,
+      const void* /*input*/,
+      const void* /*timeZone*/ = nullptr) {
+    VELOX_CHECK(
+        isSupportedType(inputTypes[0], true),
+        "to_json function does not support type {}.",
+        inputTypes[0]->toString());
+    sessionTimezone_ = getTimeZoneFromConfig(config);
+    ignoreNullFields_ = config.sparkJsonIgnoreNullFields();
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Generic<T1>>& input) {
+    return callImpl(result, input, sessionTimezone_);
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Generic<T1>>& input,
+      const arg_type<Varchar>& timeZone) {
+    auto tz = tz::locateZone(std::string_view(timeZone));
+    return callImpl(result, input, tz);
+  }
+
+ private:
+  /// Determine whether a given input type is supported.
+  /// 1. The root type can only be ROW, ARRAY, and MAP.
+  /// 2. The key type of MAP must be VARCHAR.
+  bool isSupportedType(const TypePtr& type, bool isRootType) {
+    switch (type->kind()) {
+      case TypeKind::ROW: {
+        for (const auto& child : asRowType(type)->children()) {
+          if (!isSupportedType(child, false)) {
+            return false;
+          }
+        }
+      }
+      case TypeKind::ARRAY: {
+        return isSupportedType(type->childAt(0), false);
+      }
+      case TypeKind::MAP: {
+        return type->childAt(0)->kind() == TypeKind::VARCHAR &&
```

Comment:
In Spark, we do not check the type of the map key, and your implementation also supports non-string map keys. Please double check.

Reply:
You are right. The only limit of Spark map key type is that it should not be/contain map.